### PR TITLE
LibCore: Slightly rework the Core::Promise API

### PR DESCRIPTION
**Pull request by alimpfard**
The previous iteration of this API was somewhat odd and rough in random places, which degraded usability and made less than perfect sense. This commit reworks the API to be a little closer to more conventional promise APIs (a la javascript promises).

Also adds a test to ensure the class even works.

cc @Zaggy1024.

### CODE REVIEW

File: Userland/Services/WebDriver/Session.cpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ ErrorOr<NonnullRefPtr<Core::LocalServer>> Session::create_server(NonnullRefPtr<S
         auto maybe_connection = adopt_nonnull_ref_or_enomem(new (nothrow) WebContentConnection(move(client_socket)));
         if (maybe_connection.is_error()) {
             // Use of MUST in this function is safe, as our promise callback can never error out.
```
Comment (LucasChollet) [inline Userland/Services/WebDriver/Session.cpp:68]: You can get rid of that comment

File: Userland/Libraries/LibCore/Promise.h [hunk 2]
```diff
@@ -1,76 +1,109 @@
...
+            return {};
+        };
+        on_rejection = [new_promise](ErrorType& error) {
+            new_promise->reject(move(error));
         };
         return new_promise;
     }
 
+    template<typename F>
+    requires(IsCallableWithArguments<F, ErrorOr<void>, Result&>)
```
Comment (LucasChollet) [inline Userland/Libraries/LibCore/Promise.h:77]: Is it time to introduce the CallableWithArguments concept?

File: Userland/Libraries/LibCore/Promise.h [hunk 3]
```diff
@@ -1,76 +1,109 @@
...
+    {
+        on_resolution = move(handler);
+        return *this;
+    }
+
+    template<typename F>
+    requires(IsCallableWithArguments<F, void, ErrorType&>)
+    Promise& when_rejected(F handler)
+    {
+        on_rejection = move(handler);
```
Comment (Zaggy1024) [inline Userland/Libraries/LibCore/Promise.h:88]: If the promise is running on another thread, `m_result_or_rejection` may already contain something for us to process here. Would it be possible to guarantee that `on_rejection()` is called in this function when that's the case?

File: Userland/Libraries/LibCore/Promise.h (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibCore/Promise.h:77]: Probably :P

File: Userland/Libraries/LibCore/Promise.h (hunk 3 above)
Comment (alimpfard) [reply Userland/Libraries/LibCore/Promise.h:88]: hmm...sure, can do a simple check here, mutex though...you might have to put the whole promise behind one

Comment (Zaggy1024) [reply Userland/Libraries/LibCore/Promise.h:103]: I was thinking perhaps we could take the `Optional<ErrorOr<Result, ErrorType>>` out and instead store something equivalent to an atomic-protected `Variant<Empty, Result, ErrorType>`:

```cpp
enum class ContentsType {
    Empty,
    HasResult,
    HasError,
};

Atomic<Contents> m_contents_type;
union {
    Result result;
    ErrorType error;
} m_contents;
```

We could potentially even store raw Contents and use the atomic builtin wrappers to check the contents based on a template parameter.

I'm not expert on atomics, though, so I can't say I know for sure this works as well as a mutex.

Comment (alimpfard) [reply Userland/Libraries/LibCore/Promise.h:103]: :thinking:
we could make a `Core::ThreadSafePromise` that incorporates a mutex, but what's wrong with `MutexProtected<Core::Promise<foo>>`?

**Review by Zaggy1024 — APPROVED** [review]

**Pull request merged**